### DISCUSSION TITLE
Why the mass and stiffness matrices extracted from SOFA do not match the right matrices extracted from COMSOL ?

### DISCUSSION THREAD

**Discussion by SustechShenCong** [sofa-framework/sofa#5447] [Write / Run a simulation] [answered]
In my work I need to extract the dynamical model from SOFA, i.e. extracting mass and stiffness matrices from SOFA. I follow the SOFA instructions which use "mass_matrix = root.mass.assembleMMatrix()" & "stiffness_matrix = root.force_field.assembleKMatrix()" to obtain the mass and stiffness matrices. My test example is a cube with 8 Square unit, which results 27 nodes, i.e. 81dofs. see
![mesh](https://github.com/user-attachments/assets/37529452-396a-48df-8a01-7c330d8297b3)
The COMSOL result of mass matrix is below: 
![Comsol_M](https://github.com/user-attachments/assets/ef3b9080-8bd6-44f3-9d17-da54f38b01b5)
The SOFA result of mass matrix is below: 
![SOFA_M](https://github.com/user-attachments/assets/a19ea4e4-f756-4b6a-badf-c1ec20719a59)

**Comment by SustechShenCong** [comment]
In this issue I only consider the shape of the matrix, not the detailed values!!!

**Comment by SustechShenCong** [comment]
Solution refers to https://github.com/sofa-framework/sofa/discussions/3134

**Comment by alxbilger** [comment, marked as answer]
I assume that you use MeshMatrixMass.

In MeshMatrixMass, the mass integration is distributed on vertices and edges. In a hexahedron, some vertices are not connected by edges. Therefore, this type of interaction is not taken into account in the mass matrix. And that is why, you see a lot more zeros in the matrix in SOFA.

I guess you will have more accurate results with tetrahedra, where this situation does not happen.

**Comment by SustechShenCong** [comment]
Thank you very much for your prompt reply! Your suggestions will be very helpful to me!

**Comment by SustechShenCong** [comment]
Another question, If there are some ways to extract the **Constraints Matrix** and "Internal Force Vector and its Jacobian". Some simple examples would be greatly appreciated!

**Comment by alxbilger** [comment]
There is a work in progress to extract the constraint matrix.
The internal force vector is available in the MechanicalObject, in the `force` Data field (but it is merged with all the other forces).

**Comment by SustechShenCong** [comment]
Thank you very much for your reply!
